logging module

Dominik-Vogel opened the pull request:
In this pr the logging module from the qdev_wrappers gets ported to qcodes core and its functionality gets significantly extended. For a quick overview please refer to the [example notebook](https://github.com/QCoDeS/Qcodes/blob/467e4bb6e8642d3d5184fd19fbbed8e08eaa8eb4/docs/examples/logging/logger.ipynb).

The pr is ready except for getting the CI to pass.

astafan8 commented:
@Dominik-Vogel I'm going to go through the code details soon, but could you make a brief summary of changes in the description so that I don't miss any features that you're adding/changing?

jenshnielsen commented on requirements.txt:
```diff
@@ -9,3 +9,4 @@ jsonschema
 pyzmq>=16.0.2
 broadbean>=0.9.1
 wrapt
+pandas
```
Should pandas be added to the real requirements in setup.py? Alternatively could we move the pandas import into the functions that use it so that it's only a dependency when used

Dominik-Vogel replied:
Yeah, I was also unsure about this one. @WilliamHPNielsen convinced me to have it as a general requirement.
When moving the import to the function we would also need to fix the type annotations of pandas objects used outside, or in in the header for the function...

astafan8 commented on qcodes/utils/logger.py:
```diff
@@ -0,0 +1,439 @@
+    # could be faster than this string comprehension
+
+    split_cont = [line.split(separator) for line in log
+                  if line[0].isdigit()]  # avoid tracebacks
```
perhaps mention this in the docs?

astafan8 commented on qcodes/utils/logger.py:
```diff
@@ -0,0 +1,439 @@
+        separator: seperator of the logfile to seperate the columns.
+
+    Retruns:
+        Pandas DataFrame containing the logfile content.
```
related comments about `log_to_dataframe` apply here as well.

astafan8 commented on qcodes/utils/logger.py:
```diff
@@ -0,0 +1,439 @@
+    containing time stamp strings as their values.
+
+    Args:
+        firsttimes: The oldest times
```
perhaps, rename the arguments to `oldest_times` instead of `firsttimes`? :)

astafan8 commented on qcodes/utils/logger.py:
```diff
@@ -0,0 +1,439 @@
+    if ',' in secondtimes.iloc[0]:
+        nsecondtimes = secondtimes.str.replace(',', '.')
+    else:
+        nsecondtimes = secondtimes
```
no snake_case ? :(

Dominik-Vogel replied:
@WilliamHPNielsen

astafan8 commented on qcodes/utils/logger.py:
```diff
@@ -0,0 +1,439 @@
+    original_levels = [h.level for h in handler]
+    for h in handler:
+        h.setLevel(level)
+    yield
```
missing `try`-`finally`?

astafan8 commented on qcodes/utils/logger.py:
```diff
@@ -0,0 +1,439 @@
+    """
+    global console_handler
+    with handler_level(level, handler=console_handler):
+        yield
```
missing `try`-`finally`?

Dominik-Vogel replied:
I think there is nothing to be torn down here...

astafan8 replied:
indeed!

astafan8 commented on qcodes/utils/logger.py:
```diff
@@ -0,0 +1,439 @@
+        with handler_level(level, handler):
+            yield
+    else:
+        yield
```
missing `try`-`finally`?

astafan8 commented on qcodes/utils/logger.py:
```diff
@@ -0,0 +1,439 @@
+
+
+@contextmanager
+def console_level(level: LevelType):
```
is there no convenience function for the filelog level so that the filelog always reliable logs everything (..that it is supposed to log)?

Dominik-Vogel replied:
That was the idea...

astafan8 commented on qcodes/utils/logger.py:
```diff
@@ -0,0 +1,439 @@
+    """
+    if isinstance(handler, logging.Handler):
+        handler = (handler,)
+    original_levels = [h.level for h in handler]
```
what if handler is None?

astafan8 commented on qcodes/utils/logger.py:
```diff
@@ -0,0 +1,439 @@
+
+
+@contextmanager
+def capture_dataframe(level: LevelType=logging.DEBUG,
```
is this a substitute for `LogCapture`? :) great!

astafan8 commented on qcodes/utils/logger.py:
```diff
@@ -0,0 +1,439 @@
+
+    logger.addHandler(string_handler)
+
+    yield string_handler, lambda: log_to_dataframe(
```
missing `try`-`finally`?

astafan8 commented on qcodes/utils/logger.py:
```diff
@@ -0,0 +1,439 @@
+    """
+    # get root logger if none is specified.
+    logger = logger or logging.getLogger()
+    log_capture = io.StringIO()
```
for safety, shall this be a `with ... ` statement?

astafan8 replied:
not sure if performance and memory consumption is of relevance for this particular function, but still: will the memory consumption decrease if the writing was happenning directly into the dataframe?

astafan8 commented on qcodes/utils/logger.py:
```diff
@@ -0,0 +1,439 @@
+        >>>     data_frame = cb()
+
+    Args:
+        level: level to at which to capture
```
typo: "to" or "at"?

astafan8 commented on qcodes/utils/logger.py:
```diff
@@ -0,0 +1,439 @@
+    Example:
+        >>> with logger.capture_dataframe() as (handler, cb):
+        >>>     qdac.ch01(1)  # some commands
+        >>>     data_frame = cb()
```
if one calls `cb` multiple times, will it return the complete log at the moment of the call, or the log captured between previous call and this call?

astafan8 commented:
Some comments about the notebook:
* `start_all_logging` - what is the meaning of the stuff that it prints? "logging hasn't started"?
* it would be nicer if it was more verbose. like "this context manager returns us a handler that we will pass here and there so that this and that happens", etc.
* there is no overview of the functions: like "here they are" and "this one shall be used for this, that one for that, etc"

Dominik-Vogel commented on qcodes/config/qcodesrc_schema.json:
```diff
@@ -36,26 +61,17 @@
+                "logger_levels":{
+                    "type": "object",
+                    "additionalProperties": {
+                        "type": "string"
```
Yes there is, but I wanted to keep it open in case someone wants to use user defined log levels at some point....

Dominik-Vogel commented on qcodes/instrument/base.py:
```diff
@@ -183,10 +201,10 @@ def snapshot_base(self, update: bool=False,
-                         exec_info=True)
+                self.log.warning(f"Snapshot: Could not update parameter: "
+                                 f"{name}")
+                self.log.info(f"Details for Snapshot:",
```
yes, there is unfortunately no other way. I could go through the instruments and add those lines. For now all the log messages from the `BaseInstrument` get logged.

jenshnielsen replied:
We should really have a how to write an instrument guide where things like this are mentioned. For now it might be good to add it to https://github.com/QCoDeS/Qcodes/blob/master/docs/examples/Creating%20Instrument%20Drivers.ipynb

Dominik-Vogel replied:
Fully agreed. I can add it there.

Dominik-Vogel commented:
@astafan8 for the intensive review, there is a lot of good stuff I want to add. Some answers to your questions:
 - all that `start_all_logging` prints (not logs) comes from the ipython magic. This message about the logging hadn't been started appears because we stop the history logging as a first thing to prevent an error when starting it if it had been started already.... If you find a way to read out the status you are most welcome to insert that if clause.
 - the notebook is not supposed to be a full documentation of the module, but an example of how it can be used, like all the other example notebooks we have. For the overview of the existing functions one should refer to the module itself, or rather the sphinx documentation, which I see does not get generated here. (I will look into this)
- This includes the return types and arguments, etc. 
- We changed the doc style of functions to not include types when we converted to type hints in order not to duplicate information. This includes the `Optional` keyword.
- I did not want to rewrite the log analysis notebook. I just moved it and cited it. We could of course do that but I think this has lower prio....

Dominik-Vogel commented on qcodes/instrument/channel.py:
```diff
@@ -34,12 +34,12 @@ def __init__(self,
         # want to do in the Instrument initializer
-        super().__init__(name=name, **kwargs)
-
+        self._parent = parent
```
See added comments

astafan8 replied:
which added comments? the two lines just above? i don't understand them.

Dominik-Vogel replied:
https://github.com/QCoDeS/Qcodes/pull/1309/commits/7187b25a14af750260521672f9e4c601f14ef19b

astafan8 replied:
thanks! sorry for asking 2nd time, my github didn't update properly to see the latest commits of yours.

Dominik-Vogel commented on qcodes/instrument/base.py:
```diff
@@ -17,6 +18,22 @@
+
+    def filter(self, record):
+        try:
+            return record.instrument in self.names
```
The extra keyword is something that comes from the python logging module. But I can add the comment here as well if you want.

astafan8 replied:
no no, the `extra` keyword itself is ok and is documented in python, sure. The important part is that if one wants to use an instrument adapter/filter, then one needs to pass this dict as "extra". otherwise the filter will not work, right? so, this info is missing in the docstring somewhere.

Dominik-Vogel replied:
You need to pass it to `LoggerAdapter.__init__` documented here:
https://docs.python.org/3.6/library/logging.html?highlight=loggeradapter#logging.LoggerAdapter
I added a docstring for clarity.

astafan8 replied:
great! now it is clear how to create the InstrumentLoggerAdapter so that it works with the InstrumentFilter.

btw, is it a good idea to hide this detail into the InstrumentLoggerAdapter constructor, for example, like this:
```
class InstrumentLoggerAdapter(logging.LoggerAdapter):
    def __init__(self, instrument_instance, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self.extra.update({'instrument': self.full_name})
```
(perhaps a better way is to modify kwargs before calling super.__init__)

Dominik-Vogel replied:
Your example is missing the required `extra` argument to the `LoggerAdapter.__init__`-call.

The `InstrumentLoggerAdapter` is supposed to  be used in a single place, which is the `BaseInstrument.__init__` method. I will therefore not deviate from the default way `LoggerAdapter`s are initialized.

astafan8 replied:
oh, ok, understand. i missed the part that `extra` is a required kwarg.

Dominik-Vogel replied:
Sorry for possibly misunderstanding your point:
When logging from an instrument as it is of now, using `self.log`, the records will be logged to the `InstrumentBase` logger. 
I changed the code so that every instrument now has: `self.log_base`, `self.log_visa` and `self.log`. The reason for having the extra logs around is that I discovered, that in some drivers the write_raw function gets overwritten and the visa logger is used.
I am not sure if this is the most elegant mechanism, but it mimicks what is present already.

astafan8 replied:
my point was about smth else, but let's skip it, it's ok.

your new point about "in some drivers the write_raw function gets overwritten and the visa logger is used" - very valid point. it's tricky here because one can log from the perspective of a module or from an object instance. for the case of instruments, we want to log from the perspective of the instrument instance in order to be able to filter by that instance (and add `[<instrument name>]` prefix to log messages). However, when one implements a driver and wants to use logging, the typical pattern is to declare `log = logging.getLogger(__name__)` at the top of the module. 

I don't have any suggestions at the moment... but i think that we can define whatever works as long as we stick to it and we document it in the "how to write a driver" notebook.

astafan8 commented on qcodes/instrument/channel.py:
```diff
@@ -32,14 +32,20 @@ def __init__(self,
-
+        # Naming insanity:
+        # (see https://github.com/QCoDeS/Qcodes/issues/1140 for a nice table)
+        # this has been a confusion about names. don't use name but
```
"don't use" for what?

Dominik-Vogel replied:
for anything ;-)

astafan8 replied:
but then the next line of code uses `self.name` and `parent.name`.

i'm confused. the discussion in the link tells to use short_name everywhere (so far as i understood) because it is consistent among insrtument, parameter, channel.

Dominik-Vogel replied:
yes the `self.name` is an assignment. You cannot remove it because you would break compatibility. The use of `parent.name` could be indeed improved. That part has been there before Jens introduced the concept of `short_name`. Feel free to go through the code base and remove all those.

Dominik-Vogel commented on qcodes/tests/test_helpers.py:
```diff
@@ -11,10 +11,11 @@
                                   named_repr, make_sweep, is_sequence_of,
                                   compare_dictionaries, NumpyJSONEncoder,
                                   partial_with_docstring)
+from qcodes.utils.logger import LogCapture
```
It is indeed hard to find. Can you point me to it?

astafan8 replied:
oh, you're right, it's my bad. there are no tests for this. i just assumed that it is tested here but instead it was just used by the tests of other stuff.

Dominik-Vogel commented on qcodes/utils/log_analysis.py:
```diff
@@ -5,7 +5,11 @@
+from qcodes.utils.deprecate import deprecate
 
+
+@deprecate(reason="The logging infrastructure has moved to `qcodes.utils.logger`",
```
I think this PR is not the right place to discuss whether we have been doing deprecation right in qcodes. Please open a new issue to address this and include some alternatives that you prefer from the current system.

Dominik-Vogel replied:
Possibly the backlog is a good place to prioritize this issue and allocate time for it.

astafan8 replied:
i did not mean to question the way we deprecate things in qcodes, I think it is very good :) but sorry for a stupid question, i added an item to the backlog.

Dominik-Vogel commented on qcodes/utils/logger.py:
```diff
@@ -0,0 +1,439 @@
+
+from typing import Optional, List, Union, Sequence
+
+import pandas
```
`pandas` is the only non std and non qcodes dependency in this module. Moving these functions to a new module would require exactly the same dependencies in the new module.

astafan8 replied:
Correct. and it will release the `qcodes.utils.logger` from requiring `pandas`. 

I'm thinking of `qcodes.utils.logger` module without `pandas` and smth like `qcodes.utils.logger_pandas_analysis` that does depend on `pandas`. This way the general logging features can be imported without having pandas installed. What do you think? Or is it unimportant and too much fuzz for a small benefit?

Dominik-Vogel replied:
I think we agreed on making it a dependency: https://github.com/QCoDeS/Qcodes/pull/1309#discussion_r224741658

astafan8 replied:
ok, sure.

Dominik-Vogel commented on qcodes/utils/logger.py:
```diff
@@ -0,0 +1,439 @@
+        user_path: path to the user qcodes directory
+    """
+    path = os.environ.get(QCODES_USER_PATH_ENV,
+                          os.path.join(Path.home(), '.qcodes'))
```
I am not aware of any such infrastructure. A path variable for the `qcodesrc.json` exists that nobody seems to be using.
```
    # this is for *nix people
    env_file_name = os.environ.get("QCODES_CONFIG", "")
    schema_env_file_name = env_file_name.replace(config_file_name,
                                                 schema_file_name)
```
this however points to a file not a directory so it cannot be reused.

astafan8 replied:
Exactly, i mixed the `QCODES_USER_PATH` with `QCODES_CONFIG`. Clear now, thanks!

Dominik-Vogel commented on qcodes/utils/logger.py:
```diff
@@ -0,0 +1,439 @@
+
+def _get_qcodes_user_path() -> str:
+    """
+    Get '~/.qcodes' path or if defined the path defined in the QCODES_USER_PATH
```
This is hard coded because we want to have uniformity across fridges.

Dominik-Vogel replied:
Do you have any dependencies on the logs except users opening them in an editor?
The reason for the wish for uniformity is that there should be another process, that is independent of qcodes, to read the log files. You could argue that it could also readout this location from the qcodes config. But it seems to be a little to much that it would need to implement qcodes or go through the same cascading loading of configs without qcodes.

astafan8 replied:
you're right that qcodes config is indeed not a good location for this info.

I am just striving for flexibility, so that, if necessary, a painless change is always possible. With these expectations in mind, I also hope that the process that reads the logs is configurable, hence I can tell it where to read the logs from.

So, to conclude, it is good that there is an environment variable that allows to customize this location.

Dominik-Vogel commented on qcodes/utils/logger.py:
```diff
@@ -0,0 +1,439 @@
+    if isinstance(level, int):
+        return level
+    elif isinstance(level, str):
+        return logging.getLevelName(level)
```
But seemingly also mypy is in doubt ;-)

astafan8 replied:
I checked, indeed :) and now i also understand why you needed these helper functions.

astafan8 commented on qcodes/utils/logger.py:
```diff
@@ -0,0 +1,439 @@
+    for the columns and separtor arguments is encouraged.
+
+    Args:
+        log: log content
```
of course, but I didn't mean to mention it as a type but rather than as a "structure of" that log content. But the type annotation is actually enough here.

jenshnielsen commented:
I think this is nearly ready to land. Other than the failure to build there are a few new warning raised in the tests that where not there before 

```
Failed to close name, ignored
Traceback (most recent call last):
  File "/home/travis/build/QCoDeS/Qcodes/qcodes/instrument/base.py", line 555, in close_all
    inst = cls.find_instrument(inststr)
  File "/home/travis/build/QCoDeS/Qcodes/qcodes/instrument/base.py", line 652, in find_instrument
    raise KeyError('Instrument {} has been removed'.format(name))
KeyError: 'Instrument name has been removed'
Failed to close name2, ignored
Traceback (most recent call last):
  File "/home/travis/build/QCoDeS/Qcodes/qcodes/instrument/base.py", line 555, in close_all
    inst = cls.find_instrument(inststr)
  File "/home/travis/build/QCoDeS/Qcodes/qcodes/instrument/base.py", line 652, in find_instrument
    raise KeyError('Instrument {} has been removed'.format(name))
KeyError: 'Instrument name2 has been removed'
Failed to close name3, ignored
Traceback (most recent call last):
  File "/home/travis/build/QCoDeS/Qcodes/qcodes/instrument/base.py", line 555, in close_all
    inst = cls.find_instrument(inststr)
  File "/home/travis/build/QCoDeS/Qcodes/qcodes/instrument/base.py", line 652, in find_instrument
    raise KeyError('Instrument {} has been removed'.format(name))
KeyError: 'Instrument name3 has been removed'
Failed to close name4, ignored
Traceback (most recent call last):
  File "/home/travis/build/QCoDeS/Qcodes/qcodes/instrument/base.py", line 555, in close_all
    inst = cls.find_instrument(inststr)
  File "/home/travis/build/QCoDeS/Qcodes/qcodes/instrument/base.py", line 652, in find_instrument
    raise KeyError('Instrument {} has been removed'.format(name))
KeyError: 'Instrument name4 has been removed'
```

jenshnielsen commented on docs/examples/index.rst:
```diff
@@ -38,3 +38,13 @@ Drivers
+.. toctree::
+    :glob:
+
+    ../_notebooks/logging/*
```
Should be updated to the new style as above

Dominik-Vogel commented on qcodes/logger/instrument_logger.py:
```diff
@@ -0,0 +1,120 @@
+        handler: single or sequence of handlers which to change
+    """
+    if handler is None:
+        handler = (get_console_handler(),)
```
So if tuple does this, could we not also replace the case with the instrument in the previous comment?

jenshnielsen replied:
Yes

Dominik-Vogel replied:
The above does not seem to work... (at least for my 3.6.5):
```
import logging
h = logging.StreamHandler()
tuple(h)
```
->TypeError: 'StreamHandler' object is not iterable

jenshnielsen replied:
Sorry you are right

Dominik-Vogel commented on qcodes/logger/log_analysis.py:
```diff
@@ -0,0 +1,153 @@
+                     get_log_file_name)
+
+
+def log_to_dataframe(log: List[str]=None,
```
👍

Dominik-Vogel commented on qcodes/logger/log_analysis.py:
```diff
@@ -0,0 +1,153 @@
+    return dataframe
+
+
+def logfile_to_dataframe(logfile: Optional[str]=None,
```
Here it does actually make sense, see docstring.

jenshnielsen commented:
Hopefully Azure and Travis will pass now and then this is ready to land
